### PR TITLE
Feature: Introduction of Dockerfile 

### PR DESCRIPTION
**Pull request by phoinixgrr**
#### Summary

This introduces a multi-stage Dockerfile that builds the rtcd artifact.

#### Ticket Link

https://mattermost.atlassian.net/browse/DOPS-917

### CODE REVIEW

**Review by pfltdv — APPROVED** [review]
🏆 Nice work.

**Review by streamer45 — COMMENTED** [review]
Thanks @phoinixgrr , looks great :muscle: , I have only one question:

I see that running this without any override returns some permission errors related to the log file:
```
can't open new logfile: open rtcd.log: permission denied
```

Is there anything we can do to avoid this? I assume the running user should have permissions to write logs although I understand that in reality we would probably save them to an external path (e.g. volume).

**Comment by phoinixgrr** [comment]
> Thanks @phoinixgrr , looks great 💪 , I have only one question:
> 
> I see that running this without any override returns some permission errors related to the log file:
> 
> ```
> can't open new logfile: open rtcd.log: permission denied
> ```
> 
> Is there anything we can do to avoid this? I assume the running user should have permissions to write logs although I understand that in reality we would probably save them to an external path (e.g. volume).

Hey @streamer45,

As best practice I would recommend refraining from saving logs into the container since we need the container to be stateless. What is recommended as best practise, is to push the logs in standard output. Thus they can be easily parsed by log aggregation system such as Grafana Loki. 

Does this makes sense?

**Comment by streamer45** [comment]
> As best practice I would recommend refraining from saving logs into the container since we need the container to be stateless. What is recommended as best practise, is to push the logs in standard output. Thus they can be easily parsed by log aggregation system such as Grafana Loki.
> 
> Does this makes sense?

Alright, then the suggestion is to disable file logging when running in a container. Sounds good :+1:

**Review by streamer45 — APPROVED** [review]

**Review by spirosoik — COMMENTED** [review]
minor small change @phoinixgrr . By my experience we are trying to follow the `Standard Go Project Layout` where the `Dockerfile` lives under `/build`.

A bit of context here https://github.com/golang-standards/project-layout .

An example by MM server https://github.com/mattermost/mattermost-server/tree/master/build

**Comment by phoinixgrr** [comment]
@spirosoik  I have pushed the required changes. Let me know if I got this right 

Also pushed some fixes to the `Makefile` with https://github.com/mattermost/rtcd/pull/10/commits/f6a7abe2bac7d98022246cdbbb03a9bfefcb1d83 
Thanks for Re-reviewing @streamer45  @pfltdv

File: Makefile [hunk 1]
```diff
@@ -3,7 +3,7 @@ BUILD_HASH = $(shell git rev-parse HEAD)
 LDFLAGS += -X "github.com/mattermost/rtcd/service.buildHash=$(BUILD_HASH)"
 
 ## Check go mod files consistency
-.PHONY: gomod-check
+.PHONY: gomod-check check-style golangci-lint test build clean
```
Comment (pfltdv) [inline Makefile:6]: `.PHONY: check-style` is defined at `check-style` task at line 11. Is it possible to remove that or define `.PHONY` per task?

Comment (phoinixgrr) [reply Makefile:6]: Missed that. Rebased and fixed!

**Review by streamer45 — APPROVED** [review]
Awesome :)

**Comment by agnivade** [comment]
@spirosoik - 

> minor small change @phoinixgrr . By my experience we are trying to follow the Standard Go Project Layout where the Dockerfile lives under /build.

> A bit of context here https://github.com/golang-standards/project-layout 

We are certainly not using this. This is not recommended by the Go team and some recommendations go against Go best practices.

It's obviously fine to to place Dockerfiles under `/build`. But as general guidelines, we don't follow https://github.com/golang-standards/project-layout. Our recommended project layout is here: https://developers.mattermost.com/contribute/server/style-guide/#project-layout.

**Comment by spirosoik** [comment]
@agnivade thanks for the information, I was actually mentioning only the `/build` part but thanks for sharing

**Pull request merged**